narwhal/primary/src/tests/core_tests.rs

```diff
@@ -1028,6 +1028,128 @@ async fn recover_core_expecting_header_of_previous_round() {
+        tx_certificate_waiter.clone(),
+        None,
+    );
+
+    let metrics = Arc::new(PrimaryMetrics::new(&Registry::new()));
+
+    let own_address = network::multiaddr_to_address(&committee.primary(&name).unwrap()).unwrap();
+    println!("New primary added: {:?}", own_address);
+    let network = anemo::Network::bind(own_address)
+        .server_name("narwhal")
+        .private_key(primary.network_keypair().copy().private().0.to_bytes())
+        .start(anemo::Router::new())
+        .unwrap();
+
+    let mut primary_networks = Vec::new();
+    let fixture_authorities = fixture.authorities().filter(|a| a.public_key() != name);
+    for primary in fixture_authorities {
+        let address = committee.primary(&primary.public_key()).unwrap();
+
+        let mut mock_server = MockPrimaryToPrimary::new();
+        mock_server
+            .expect_get_latest_header()
+            .returning(move |_request| {
+                let builder = types::HeaderBuilder::default();
+                let fixture = CommitteeFixture::builder().randomize_ports(true).build();
+                let primary = fixture.authorities().next().unwrap();
+
+                let header = builder
+                    .author(primary.public_key())
+                    .round(1)
+                    .epoch(0)
+                    .parents([CertificateDigest::default()].iter().cloned().collect())
+                    .with_payload_batch(fixture_batch_with_transactions(10), 0)
+                    .build(primary.keypair())
+                    .unwrap();
+                Ok(anemo::Response::new(LatestHeaderResponse {
+                    header: Some(header),
+                }))
+            });
+
+        let routes =
+            anemo::Router::new().add_rpc_service(types::PrimaryToPrimaryServer::new(mock_server));
+        primary_networks.push(primary.new_network(routes));
+
+        let address = network::multiaddr_to_address(&address).unwrap();
+        let peer_id = PeerId(primary.network_keypair().public().0.to_bytes());
+        network
+            .connect_with_peer_id(address, peer_id)
+            .await
+            .unwrap();
+    }
+
+    // Spawn the core.
+    let _core_handle = Core::spawn(
+        name.clone(),
+        committee.clone(),
+        worker_cache.clone(),
+        header_store.clone(),
+        certificates_store.clone(),
+        create_test_vote_store(),
+        synchronizer,
+        signature_service.clone(),
+        rx_consensus_round_updates.clone(),
+        /* gc_depth */ 50,
+        rx_reconfigure.clone(),
+        /* rx_primaries */ rx_primary_messages,
+        /* rx_header_waiter */ rx_headers_loopback,
+        /* rx_certificate_waiter */ rx_certificates_loopback,
+        /* rx_proposer */ rx_headers,
+        tx_consensus,
+        /* tx_proposer */ tx_parents,
+        metrics.clone(),
+        P2pNetwork::new(network.clone()),
+    );
+
+    tokio::time::sleep(Duration::from_secs(5)).await;
```

Comment:
instead of a 5 second sleep which is a bit of a kludge, if you want to wait for your mock handler to be called before proceeding, then use a oneshot that gets signaled from inside the closure. for an example see `worker_tests.rs:72`

Reply:
Because this is used in conjunction with paused time via `#[tokio::test(flavor = "current_thread", start_paused = true)]`, it doesn't case the test to actually sleep for 5 seconds of wall time

Reply:
I don't think that obviates my suggestion? There is no guarantee that sleeping for 5 seconds actually awaits the thing you want to await, whereas if you synchronize it with a channel there is.

Reply:
Ah, I see what your concern was. Unfortunately, the mock handler being called doesn't guarantee that the header has been processed yet either, and is likely to be premature, since we are now collecting the headers in a vector and calling process header on all after all the responses are received, which is also done via async fn call...

Reply:
In that case could you notify_read on the Header store? That would actually be better than using the metric to verify behavior anyway, if the processing in your test gets that far. If that won't work, I guess this test should be deleted again soon...